src/lib/vasi-sync/src/lazy_lock.rs

```diff
@@ -0,0 +1,176 @@
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::sync::atomic::{self, AtomicI8};
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use vasi::VirtualAddressSpaceIndependent;
+
+use crate::sync::{ConstPtr, UnsafeCell};
+
+// For use with `LazyLock`. This trait is implemented for
+// `FnOnce` closures, so most users can just use that directly.
+// However in case they need to name a type in the LazyLock,
+// they can create a type that implements this trait instead.
+pub trait Producer<T> {
+    fn initialize(self) -> T;
+}
+
+impl<F, T> Producer<T> for F
+where
+    F: FnOnce() -> T,
+{
+    fn initialize(self) -> T {
+        (self)()
+    }
+}
+
+#[derive(TryFromPrimitive, IntoPrimitive, Eq, PartialEq)]
+#[repr(i8)]
+enum InitState {
+    Uninitd,
+    Initializing,
+    Initd,
+}
+
+/// Analogous to `std::sync::LazyLock`, but works on stable Rust, is no_std,
+/// only makes direct syscalls, etc.
+pub struct LazyLock<T, Init = fn() -> T> {
```

Comment:
This needs to be `#[repr(C)]` if it is supposed to be `VirtualAddressSpaceIndependent`. I think the vasi-sync types also need to be made `#[repr(C)]`, such as `UnsafeCell`.